src/resolve-info.ts

```diff
@@ -196,18 +231,31 @@ const memoizedExpandFieldNode = MEMOIZATION
 function expandFieldNode(
   schema: GraphQLSchema,
   fragments: FragmentsType,
-  node: FieldNode,
+  node: JitFieldNode,
   fieldType: GraphQLOutputType
 ): FieldExpansion | LeafField {
+  const shouldInclude = (variables: ShouldIncludeVariables): boolean => {
+    const expr = node.__internalShouldInclude as string;
+    // eslint-disable-next-line no-new-func
+    const fn = new Function(`__context`, `return ${expr}`);
+
+    return fn(variables);
+  };
+
   if (node.selectionSet == null) {
-    return createLeafField({});
+    return createLeafField({}, shouldInclude);
   }
 
   // there is a selectionSet which makes the fieldType a CompositeType
   const typ = memoizedResolveEndType(fieldType) as GraphQLCompositeType;
   const possibleTypes = memoizedGetPossibleTypes(schema, typ);
 
-  const fieldExpansion: FieldExpansion = {};
+  const fieldExpansion: FieldExpansion = Object.create(
+    {
+      __shouldInclude: shouldInclude
+    },
+    {}
+  );
```

Comment:
I have 0 ideas how, but these lines break objects merging and lead to the failing test in this PR.

Reply:
The object merging breaks because we set here a prototype for the object, but lodash `mergeWith` is expecting objects to be plain objects to function, and it does not know how to merge objects that have non `null` prototypes.

See https://lodash.com/docs/4.17.15#merge

> Array and plain object properties are merged recursively. Other objects and value types are overridden by assignment.

Basically because we are not having plain objects anymore, in this code https://github.com/lodash/lodash/blob/2da024c3b4f9947a48517639de7560457cd4ec6c/.internal/baseMergeDeep.js#L69-L89 we skip this merge part, and just override the value, as stated above

Reply:
Our `customizer` for `mergeWith` is only customizing the leaf nodes, and returns `undefined` for other nodes, i.e. then normal lodash merge is used.

One way to deal with this would be to extend the customizer function to also merge our new objects that have prototype https://github.com/zalando-incubator/graphql-jit/blob/9de816e29801eb692f51993ecdfeb171a6d44ae8/src/resolve-info.ts#L378-L393